### DISCUSSION TITLE
with_inner_size does not set a PhysicalSize for the window

### DISCUSSION THREAD

**Discussion by devrangel** [rust-windowing/winit#3262] [Q&A] [answered]
I'm trying to set a initial size for the window, but `inner_size()` is returning (0, 0).

```rust
let window = WindowBuilder::new()
  .with_title("Example")
  .with_inner_size(PhysicalSize::new(100, 100))
  .with_append(true)
  .build(&event_loop)
  .unwrap();

let size = window.inner_size(); // return (0, 0)
info!("{} {}", size.width, size.height)

web_sys::window()
  .and_then(|win| win.document())
  .and_then(|doc| {
      let canvas = window.canvas().unwrap();
  
      let body = doc.body().unwrap();
      let body_style = &body.style();
      body_style.set_property("height", "100vh");
  
      let style = &canvas.style();
      style.set_property("display", "block");
      style.set_property("width", "100%");
      style.set_property("height", "100%");
  
      canvas.set_width(canvas.offset_width() as u32);
      canvas.set_height(canvas.offset_height() as u32);
  
      let body = doc.body().unwrap();
      body.append_child(&canvas).ok()?;
  
      Some(())
  })
  .expect("Could't initialize!");
```
The window only gets its values after the `WindowEvent::Resized` event.

```rust
WindowEvent::Resized(physical_size) => state.resize(*physical_size),
```
But I need the window size before the Resized event in order to configure the wgpu.

```rust
let config = wgpu::SurfaceConfiguration {
  usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
  format: surface_format,
  width: size.width,
  height: size.height,
  present_mode: surface_caps.present_modes[0],
  alpha_mode: surface_caps.alpha_modes[0],
  view_formats: vec![]
};
```
`Target: wasm32`
`wgpu = "0.18.0"`
`winit = { version = "0.29.4", features = ["rwh_05"] }`

**Comment by daxpedda** [comment, marked as answer]
See https://github.com/rust-windowing/winit/issues/2863.

Unfortunately this is not possible, as we use `ResizeObserver` to figure out the size of the canvas, which is async by nature. This is unfortunately a limitation by the Web API.

**Reply by devrangel** [reply]
`WindowEvent::Resized` event is called at least once after initialization, right?

I thought about passing the width and height of the canvas, multiplied by the monitor scale factor, as arguments to the wgpu initialization method. And after initialization, WindowEvent::Resized is called and assigns the final size to the Window.

Is it a good approach?

**Reply by daxpedda** [reply]
> `WindowEvent::Resized` event is called at least once after initialization, right?

Yes.

> I thought about passing the width and height of the canvas, multiplied by the monitor scale factor, as arguments to the wgpu initialization method. And after initialization, WindowEvent::Resized is called and assigns the final size to the Window.

Don't use the canvas `width` or `height`, this is why this whole thing turned out so problematic, as those measurements don't actually report the size of the canvas but the size of the output bitmap. You can use [getComputedStyle()](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle), which should do the job.

The `WindowEvent::Resized` that will happen later should at most be one pixel off.

> Is it a good approach?

The recommended approach would be to initialize Wgpu with any size, but not start drawing until you receive `WindowEvent::Resized` and properly configure the surface with the right size.

EDIT: I don't know when you start drawing in your approach, but it seems fine to me as well.
